pallets/foreign-investments/src/impls/mod.rs

```diff
@@ -0,0 +1,1152 @@
+				invest_swap_amount.ensure_add(redeem_swap_amount)?,
+				redeem_state,
+			)),
+		}
+		.map(|(token_swap_amount, maybe_new_redeem_state)| {
+			// If old state match new state, no need to return it as this could cause a
+			// follow-up transition trigger
+			if redeem_state == maybe_new_redeem_state {
+				(token_swap_amount, None)
+			} else {
+				(token_swap_amount, Some(maybe_new_redeem_state))
+			}
+		})
+		.map_err(|e: DispatchError| e)?;
+
+		// Determine token swap from amount
+		let token_swap = if invest_swap_amount > redeem_swap_amount {
+			invest_state.get_active_swap().map(|invest_swap| Swap {
+				amount: final_swap_amount,
+				..invest_swap
+			})
+		}
+		// handle redeem_swap_amount >= invest_swap_amount as well as all cases, in which neither
+		// states include an active swap
+		else {
+			redeem_state.get_active_swap().map(|redeem_swap| Swap {
+				amount: final_swap_amount,
+				..redeem_swap
+			})
+		};
+
+		Ok((token_swap, new_invest_state, new_redeem_state))
+	}
+
+	/// Increments the collected investment amount and transitions investment
+	/// state as a result of collecting the investment.
+	///
+	/// NOTE: Does not transfer back the collected tranche tokens. This happens
+	/// in `transfer_collected_investment`.
+	pub(crate) fn denote_collected_investment(
+		who: &T::AccountId,
+		investment_id: T::InvestmentId,
+		collected: CollectedAmount<T::Balance>,
+	) -> DispatchResult {
+		// Increment by previously stored amounts (via `CollectedInvestmentHook`)
+		CollectedInvestment::<T>::mutate(who, investment_id, |collected_before| {
+			collected_before
+				.amount_collected
+				.ensure_add_assign(collected.amount_collected)?;
+			collected_before
+				.amount_payment
+				.ensure_add_assign(collected.amount_payment)?;
+			Ok::<(), DispatchError>(())
+		})?;
+
+		// Update invest state to decrease the unprocessed investing amount
+		let investing_amount = T::Investment::investment(who, investment_id)?;
+		let pre_state = InvestmentState::<T>::get(who, investment_id);
+		let post_state =
+			pre_state.transition(InvestTransition::CollectInvestment(investing_amount))?;
+		Self::apply_invest_state_transition(who, investment_id, post_state).map_err(|e| {
+			log::debug!("InvestState transition error: {:?}", e);
+			Error::<T>::from(InvestError::Collect)
+		})?;
+
+		Ok(())
+	}
+
+	/// Consumes the `CollectedInvestment` amounts and returns these.
+	///
+	/// NOTE: Converts the collected pool currency payment amount to foreign
+	/// currency via the `CurrencyConverter` trait.
+	pub(crate) fn transfer_collected_investment(
+		who: &T::AccountId,
+		investment_id: T::InvestmentId,
+		foreign_payout_currency: T::CurrencyId,
+		pool_currency: T::CurrencyId,
+	) -> Result<ExecutedForeignCollectInvest<T::Balance>, DispatchError> {
+		let collected = CollectedInvestment::<T>::take(who, investment_id);
+		ensure!(
+			!collected.amount_payment.is_zero(),
+			Error::<T>::InvestError(InvestError::NothingCollected)
+		);
```

Comment:
Why this can not be zero? What happens if the user collects and they didn't know it wasn't ready to be collected?

Reply:
When a user collects, we send back the `ExecutedCollectInvest` message to the sending domain. IMO, we only want to do this if the amount is not zero. Else the frontend should propagate the error and explain the user no collection could be made at this point. 

WDYT @mustermeiszer?

Reply:
Then, maybe the protection to avoid sending the message should be in `pallet-liquidity-pools`? Anyways, I feel like collecting 0 things is a correct behavior. The UI should show something like "collected: 0". But not sure if an error.

If I'm adding noise with this, please skip/avoid this thread